src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java

```diff
@@ -513,4 +515,89 @@ public void evaluate() throws Throwable {
 			}
 		});
 	}
+
+	@Issue("JENKINS-40879")
+	@Test
+	public void parallelLockRelease() throws Exception {
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				LockableResourcesManager.get().createResource("resource1");
+				LockableResourcesManager.get().createResource("resource2");
+				LockableResourcesManager.get().createResource("resource3");
+				LockableResourcesManager.get().createResource("resource4");
+				WorkflowJob j = story.j.jenkins.createProject(WorkflowJob.class, "j");
+				j.setDefinition(new CpsFlowDefinition(
+						"lock(resource: 'resource1') {\n" +
+								"    sleep 1\n" +
```

Comment:
> we do want the other runs to still be in the `lock`s when we take one run out of the last `lock`

No problem—that is why you interleave `waitForStart` and `success` calls between builds.

> we don't want to dictate exactly when/which run goes for the last lock next

Sure you do. If you understand the scenario under which the error happens, then you can orchestrate it precisely.